src/accounts.rs

```diff
@@ -250,42 +527,74 @@ impl AccountsDB {
             .collect()
     }
 
-    pub fn increment_transaction_count(&mut self, tx_count: usize) {
-        self.transaction_count += tx_count as u64
+    pub fn increment_transaction_count(&self, tx_count: usize) {
+        let mut tx = self.transaction_count.write().unwrap();
+        *tx += tx_count as u64;
     }
 
     pub fn transaction_count(&self) -> u64 {
-        self.transaction_count
+        let tx = self.transaction_count.read().unwrap();
+        *tx
     }
     pub fn account_values_slow(&self) -> Vec<(Pubkey, solana_sdk::account::Account)> {
-        self.accounts.iter().map(|(x, y)| (*x, y.clone())).collect()
+        let mut vec: Vec<(Pubkey, Account)> = vec![];
+        let fork: Fork = 0;
+        let rindex = self.index_info.index.read().unwrap();
+        let stores = self.storage.read().unwrap();
+        rindex.iter().for_each(|(p, x)| {
+            if let Some((id, index)) = x.0.get(&fork) {
+                let account = stores[*id]
+                    .appendvec
+                    .read()
+                    .unwrap()
+                    .get_account(*index)
+                    .unwrap();
+                vec.push((*p, account));
+            }
+        });
+        vec
     }
-    fn merge(&mut self, other: Self) {
-        self.transaction_count += other.transaction_count;
-        self.accounts.extend(other.accounts)
+    fn merge(&self, other: &Self) {
+        let mut wtx = self.transaction_count.write().unwrap();
+        let rtx = other.transaction_count.read().unwrap();
+        *wtx += *rtx;
+        let mut stores = self.storage.write().unwrap();
+        let prev_len = stores.len();
+        stores.append(&mut other.storage.write().unwrap());
+        let rindex = other.index_info.index.read().unwrap();
+        let mut windex = self.index_info.index.write().unwrap();
+        rindex.iter().for_each(|(p, x)| {
+            let forks = windex.entry(*p).or_insert(AccountMap(HashMap::new()));
+            x.0.iter().for_each(|(fork, (id, index))| {
+                forks.0.insert(*fork, (*id + prev_len, *index));
+            });
+        });
     }
 }
 
 impl Accounts {
+    #[cfg(test)]
+    pub fn new(paths: &[String]) -> Self {
+        Accounts {
+            accounts_db: AccountsDB::new(&paths),
+            account_locks: Mutex::new(HashSet::new()),
+        }
+    }
+
     /// Slow because lock is held for 1 operation insted of many
     pub fn load_slow<U>(checkpoints: &[U], pubkey: &Pubkey) -> Option<Account>
```

Comment:
@sambley so all this checkpoints stuff can be removed, it just needs to take the fork id